### PR TITLE
[refactor] Split constructing and compilation of lang::Function

### PR DESCRIPTION
**Pull request by PGZXB**
Issue: #7002 

### Brief Summary
Removed dependencies on `Program::this_thread_config()` in `lang::Function` compilation (AST->IR part)
* Push off the compilation of `lang::Function`: Introduce the `irpass::compile_called_function(IRNode *root, const CompileConfig &config)`, which compiles the AST/IR of `Function`s called in `root` to the final IR.

### CODE REVIEW

File: taichi/program/function.h [hunk 1]
```diff
@@ -9,6 +9,8 @@ class Program;
 
 class Function : public Callable {
  public:
+  enum class IRType { None, AST, InitialIR, OptimizedIR };
```
Comment (ailzhang) [inline taichi/program/function.h:12]: nit: might need better names here. cc: @strongoier

File: taichi/ir/transforms.h [hunk 2]
```diff
@@ -199,6 +199,8 @@ void compile_function(IRNode *ir,
 void ast_to_ir(const CompileConfig &config,
                Kernel &kernel,
                bool to_executable = true);
+
+void compile_called_functions(IRNode *ir, const CompileConfig &compile_config);
```
Comment (ailzhang) [inline taichi/ir/transforms.h:203]: nit: `compile_called_function` isn't very intuitive. how about `compile_taichi_function`?

File: taichi/program/function.h (hunk 1 above)
Comment (strongoier) [reply taichi/program/function.h:12]: It seems OK for now :-)

File: taichi/transforms/compile_called_functions.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+ private:
+  const CompileConfig &compile_config_;
+};
+
+namespace irpass {
+
+void compile_taichi_functions(IRNode *ir, const CompileConfig &compile_config) {
+  TI_AUTO_PROF;
+  CompileCalledFunctions::run(ir, compile_config);
```
Comment (strongoier) [inline taichi/transforms/compile_called_functions.cpp:45]: nit: name mismatch?

**Review by ailzhang — APPROVED** [review]
Thanks!

**Pull request merged**